### PR TITLE
Closes #2686: `categorical_test.py` conversion for new test framework

### LINKED ISSUE #2686: `categorical_test.py` conversion for new test framework
Convert `categorical_test.py` to the new test framework

### PR DESCRIPTION
**Pull request by pierce314159**
This PR (closes #2686) converts `categorical_test.py` to new test framework.

The io tests that were not yet covered in `io_test.py` were added. `test_deletion` is already covered in `symbol_table_test.py`

### CODE REVIEW

File: PROTO_tests/tests/categorical_test.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            ak.Categorical(ak.arange(0, 5, 10))
+
+    def test_from_codes(self):
+        codes = ak.array([7, 5, 9, 8, 2, 1, 4, 0, 3, 6])
+        categories = ak.array([f"string {i}" for i in range(10)] + ["N/A"])
+
+        cat = ak.Categorical.from_codes(codes, categories)
+        assert codes.to_list() == cat.codes.to_list()
+        assert categories.to_list() == cat.categories.to_list()
+
```
Comment (TheOddCzar789) [inline PROTO_tests/tests/categorical_test.py:46]: Do we want to do a TypeError check for `test_from_codes`?

Comment (pierce314159) [reply PROTO_tests/tests/categorical_test.py:46]: nah I don't think it's necessary, the argument validation is pretty straightforward for this method

**Review by TheOddCzar789 — APPROVED** [review]

File: PROTO_tests/tests/categorical_test.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    def test_substring_search(self):
+        cat = self.create_basic_categorical()
+        # update to use regex flag once #2714 is resolved
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/categorical_test.py:49]: Noting that the PR for this issue has been approved by at least 1 person, please be sure to update the appropriate code based on which gets merged first.

**Review by Ethan-DeBandi99 — APPROVED** [review]
I don't think that this is a test where we will be able to parameterize much easily, so I think this is good to go.

File: PROTO_tests/tests/categorical_test.py (hunk 2 above)
Comment (Ethan-DeBandi99) [reply PROTO_tests/tests/categorical_test.py:49]: I am going to merge this, I will not in the other PR

**Pull request merged**